content/en/blog/_posts/2021-08-11-memory-qos-cgroups-v2.md

```diff
@@ -0,0 +1,73 @@
+---
+layout: blog
+title: 'Support memory qos with cgroups v2'
+date: 2021-08-11
+slug: memory-qos-cgroups-v2
+---
+
+**Authors:** Tim Xu (Tencent Cloud)
+
+Memory QoS is used to memory protection and usage throttle on pod / container to increase memory isolation and service quality.
+
+In traditional cgroups v1 in Kubernetes, cpu resources can be very well protected. Kubernetes uses `cpu_shares / cpu_set / cpu_quota / cpu_period` to isolate cpu. But the memory is not well guaranteed, especially at cgroup level. Kubernetes only uses `memory.limit_in_bytes=sum(pod.spec.containers.resources.limits[memory])` with cgroups v1 and `memory.max=sum(pod.spec.containers.resources.limits[memory])` with cgroups v2 to limit memory usage. `resources.requests[memory]` is not yet used neither by cgroups v1 nor cgroups v2 to protect memory requests. About memory protection, we use `oom_scores` to determine order of killing container process when OOM occurs.
+
+It may cause:
+- Pod / Container memory requests can't be fully reserved, page cache is at risk of being recycled
```

Comment:
```suggestion
- Pod or Container memory requests to not be fully reserved and page cache to become at risk of being recycled
```

Reply:
I'd write:
```suggestion
- The kubelet may not be able to guarantee memory resources even when you
  specify these with a `limit` or a `request` for a Pod or for one of its containers.
  Because the memory could be overcommitted at the host operating system level,
  memory pages  that are backed by non-volatile storage (such as the executable
  code for your app) could be dropped from RAM due to overall memory pressure,
   even when your Pod is running with Guaranteed QoS.
```
and I think that's also accurate technically.

Reply:
OK, Have updated.